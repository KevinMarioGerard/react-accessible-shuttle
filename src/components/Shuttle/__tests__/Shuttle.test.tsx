import * as React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';

// import '@testing-library/jest-dom/extend-expect';

import { Shuttle, useShuttleState } from '../../../index';

const dummyState = {
    source: new Array(5).fill(null).map(() => Math.round(Math.random() * 1000)),
    target: new Array(5).fill(null).map(() => Math.round(Math.random() * 1000)),
};

const TestShuttle = ({ state }: any) => {
    const shuttle = useShuttleState(state);

    return (
        <Shuttle {...shuttle} enableUserSelectionHack>
            <Shuttle.Container>
                {({ source, selected }, getItemProps) =>
                    source.map((item, index) => (
                        <Shuttle.Item
                            {...getItemProps(index)}
                            key={item}
                            value={item}
                            selected={selected.source.has(index)}>
                            {item}
                        </Shuttle.Item>
                    ))
                }
            </Shuttle.Container>
            <Shuttle.Controls data-testid="controls" />
            <Shuttle.Container>
                {({ target, selected }, getItemProps) =>
                    target.map((item, index) => (
                        <Shuttle.Item
                            {...getItemProps(index)}
                            key={item}
                            value={item}
                            selected={selected.target.has(index)}>
                            {item}
                        </Shuttle.Item>
                    ))
                }
            </Shuttle.Container>
        </Shuttle>
    );
};

afterEach(cleanup);

describe('Shuttle tests', () => {
    it('should render', () => {
        expect(() => {
            render(<TestShuttle state={dummyState} />);
        }).not.toThrow();
    });

    it('should throw when source selections or target selections is not an array', () => {
        expect(() => {
            render(<TestShuttle state={{ selections: { source: 0 } }} />);
        }).toThrow();

        expect(() => {
            render(<TestShuttle state={{ selections: { source: [], target: 0 } }} />);
        }).toThrow();
    });

    it('should select items', () => {
        const { container, getByTestId } = render(<TestShuttle state={dummyState} />);

        const button = getByTestId("controls").querySelector('button');

        if (button) {
            fireEvent.click(button);
        }

        expect(container).toMatchSnapshot();
    });
});
