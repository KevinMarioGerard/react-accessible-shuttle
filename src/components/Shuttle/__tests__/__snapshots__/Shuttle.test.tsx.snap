// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shuttle tests Shuttle smoke test and other basic operations should select items 1`] = `
<div>
  <div
    class="shuttle shuttle--ush"
    role="presentation"
  >
    <div
      class="shuttle__container"
      data-name="source"
      data-testid="source_container"
      role="listbox"
      tabindex="0"
    >
      <div
        class="shuttle__item shuttle__item--selected"
        data-index="0"
        data-value="0"
        role="option"
        tabindex="-1"
      >
        0
      </div>
      <div
        class="shuttle__item"
        data-index="1"
        data-value="1"
        role="option"
        tabindex="-1"
      >
        1
      </div>
      <div
        class="shuttle__item"
        data-index="2"
        data-value="2"
        role="option"
        tabindex="-1"
      >
        2
      </div>
      <div
        class="shuttle__item"
        data-index="3"
        data-value="3"
        role="option"
        tabindex="-1"
      >
        3
      </div>
      <div
        class="shuttle__item"
        data-index="4"
        data-value="4"
        role="option"
        tabindex="-1"
      >
        4
      </div>
    </div>
    <div
      class="shuttle__controls"
      data-testid="controls"
    >
      <button
        aria-label="Move All Items to Target"
        title="Move All Items to Target"
      >
        »
      </button>
      <button
        aria-label="Move Selected Items to Target (1 items selected in source container)"
        title="Move Selected Items to Target"
      >
        ›
      </button>
      <button
        aria-label="Move Selected Items to Target (0 items selected in target container)"
        title="Move Selected Items to Source"
      >
        ‹
      </button>
      <button
        aria-label="Move All Items to Source"
        title="Move All Items to Source"
      >
        «
      </button>
    </div>
    <div
      class="shuttle__container"
      data-name="target"
      data-testid="target_container"
      role="listbox"
      tabindex="0"
    >
      <div
        class="shuttle__item"
        data-index="0"
        data-value="0"
        role="option"
        tabindex="-1"
      >
        0
      </div>
      <div
        class="shuttle__item"
        data-index="1"
        data-value="1"
        role="option"
        tabindex="-1"
      >
        1
      </div>
      <div
        class="shuttle__item"
        data-index="2"
        data-value="2"
        role="option"
        tabindex="-1"
      >
        2
      </div>
      <div
        class="shuttle__item"
        data-index="3"
        data-value="3"
        role="option"
        tabindex="-1"
      >
        3
      </div>
      <div
        class="shuttle__item"
        data-index="4"
        data-value="4"
        role="option"
        tabindex="-1"
      >
        4
      </div>
    </div>
  </div>
</div>
`;
